### PR TITLE
Add support for "host components"

### LINKED ISSUE #311: refactor: Host components
Currently, host-implemented interfaces have to be hard-coded in `engine` in a couple of places:
- the implementation state is included in `RuntimeContext`: https://github.com/fermyon/spin/blob/2be403465390f1a767181f16190aa4f14512c9f3/crates/engine/src/lib.rs#L49-L50
- a `link_*` method, which is called in `build_default`: https://github.com/fermyon/spin/blob/2be403465390f1a767181f16190aa4f14512c9f3/crates/engine/src/lib.rs#L106-L108

I'd like to invert this dependency by introducing a new `HostComponent` trait. After a little prototyping I think it can look something like:
```rust
pub trait HostComponent {
    type State: Default + 'static;
    fn add_to_linker<T>(id: String, linker: &mut wasmtime::Linker<crate::RuntimeContext<T>>);
    fn initial_state(id: String) -> State;
}
```
implemented like:
```rust
impl HostComponent for OutboundHttp {
    type State = Self;
    fn add_to_linker<T>(id: String, linker: &mut Linker<RuntimeContext<T>>) {
        wasi_outbound_http::add_to_linker(linker, |ctx| ctx.host_components.get_state(&id).unwrap())
    }
}
```
_Note: The separate `State` type isn't necessary here but is needed for interfaces that use reftypes, which changes their state data type._

The component state would be held in a mapping on `RuntimeContext` as e.g.:
```rust
/// Host components data,
pub host_components: HashMap<String, Box<dyn Any>>,
```

There might be a performance issue here as this HashMap would be referenced once for every host component function call, but if that is the case we could switch to a more efficient form of mapping (perhaps a Vec with pre-computed indexes or even something more exotic/unsafe :grimacing:). If performance is a real problem here I _think_ its a solvable one but its a real tradeoff vs the organizational benefits of splitting things up like this. Other ideas for how to do this very welcome.

### PR DESCRIPTION
**Pull request by lann**
Host components are a way to add host implementations of component
interfaces with a dependency injection pattern. Host components
implement the HostComponent trait which defines how the host
implementation is linked and how the component's state is constructed.

Fixes #311

### CODE REVIEW

**Comment by lann** [comment]
Keeping this a draft until dependency #382 merges.

**Comment by lann** [comment]
I only updated outbound-redis and object-store in this PR to give an idea of what  `HostComponent` implementations look like. I think outbound-http will be a little more complicated to convert (but possible). I'm not sure about config yet; it might need to be kept separate if `HostComponents` end up using the config mechanism themselves. Wasi will stay in engine for now.

File: src/commands/up.rs
Comment (itowlson) [inline src/commands/up.rs:173]: This is really nice.

File: crates/outbound-redis/src/lib.rs
Comment (itowlson) [inline crates/outbound-redis/src/lib.rs:14]: I don't quite get why this would be `Self` rather than `()`.  Is the 'runtime state' always an instance of the component plus optional something else (config?)?  If so, would it be clearer to distinguish these, so that the Self is baked in and only the additional data needs to be in the `Data` type?

File: crates/object-store/src/file.rs
Comment (itowlson) [inline crates/object-store/src/file.rs:41]: Given this deconstructs and immediately reconstructs the tuple, why not just `move |ctx| data_handle.get_mut(ctx)`?

File: crates/engine/src/host_component.rs
Comment (itowlson) [inline crates/engine/src/host_component.rs:20]: `data` is such an open-ended term - I found it hard to understand what this member did at the point of implementation.  Can we be more specific e.g. `build_runtime_state`?

File: crates/engine/src/host_component.rs
Comment (itowlson) [inline crates/engine/src/host_component.rs:13]: Not sure I understand the reference to a closure - does this refer to a previous implementation?

File: crates/engine/src/host_component.rs
Comment (itowlson) [inline crates/engine/src/host_component.rs:64]: This is an index into a specific `HostComponentsData`.  Is there a way to bundle this up so they travel together instead of having this slightly orphaned-looking index?  (Is that what the closure used to do?)

**Review by itowlson — APPROVED** [review]
I love the structure of this - I love that each component defines its own linkage and that the top level now injects what it wishes.  I struggled a bit to understand the implementation.  But from a usage point of view this is great.

File: crates/outbound-redis/src/lib.rs
Comment (lann) [reply crates/outbound-redis/src/lib.rs:14]: The closure passed to `add_to_linker` just below has to return a `&mut Data`; you have to have an instance of `Data` to borrow from even if its a unit struct (unless you want to do some `transmute` jiggery-pokery, but if your `Data` has state you'll need the ref anyway).

`Data` is just the type that implements the interface _unless_ that interface uses resources, in which case `Data` is `(Impl, <InterfaceName>Tables<Impl>)` (the "tables" track resource state).

I would be happy to have a way to simplify this, but I spent way too much time trying and not enough time succeeding.

Comment (lann) [reply crates/outbound-redis/src/lib.rs:14]: To clarify: "has to return a `&mut Data`" really means "`Data` has to match the wit-bindgen-generated signature of that closure".

File: crates/object-store/src/file.rs
Comment (lann) [reply crates/object-store/src/file.rs:41]: `get_mut` returns e.g. `&mut (A, B)`, but this closure needs (&mut A, &mut B). I also spent an unreasonable amount of time trying to fix this with no luck. Suggests welcome. If they work.

File: crates/engine/src/host_component.rs
Comment (lann) [reply crates/engine/src/host_component.rs:13]: Woops! Yes indeed; I originally tried to pass a closure in rather than this "handle". I still don't understand why that wasn't essentially equivalent to this, but after much lifetime wrangling I gave up on it.

File: crates/engine/src/host_component.rs
Comment (lann) [reply crates/engine/src/host_component.rs:64]: Maybe? This handle is created well before the data, so it can't reference it directly.

File: crates/object-store/src/file.rs
Comment (itowlson) [reply crates/object-store/src/file.rs:41]: Aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaah.  That's cool - I just completely missed that nuance.

File: crates/engine/src/host_component.rs
Comment (radu-matei) [reply crates/engine/src/host_component.rs:13]: Interesting, I would have expected to be able to use a closure similar to the one in the generated Wasmtime code.

File: crates/engine/src/host_component.rs
Comment (radu-matei) [reply crates/engine/src/host_component.rs:20]: `data` is the term used by `wit-bindgen` to refer to the runtime state, so the name does make sense.

File: crates/engine/src/host_component.rs
Comment (radu-matei) [inline crates/engine/src/host_component.rs:17]: nit: could you just `use anyhow::Result` here?

File: crates/outbound-redis/src/lib.rs
Comment (radu-matei) [inline crates/outbound-redis/src/lib.rs:18]: nit: would you mind importing `Linker` and `HostComponentsDataHandle`?

**Review by radu-matei — APPROVED** [review]

File: crates/engine/src/host_component.rs
Comment (lann) [reply crates/engine/src/host_component.rs:13]: If someone can make it work I'd be happy with the change :shrug:

Comment (radu-matei) [reply crates/engine/src/host_component.rs:13]: Hah, no, I definitely wouldn't block this PR.

**Pull request merged**
